tensorflow/core/kernels/data/sampling_dataset_op.cc

```diff
@@ -0,0 +1,202 @@
+    }
+
+    ~Dataset() override { input_->Unref(); }
+
+    std::unique_ptr<IteratorBase> MakeIteratorInternal(
+        const string& prefix) const override {
+      if (rate_ == 0) {
+        return std::unique_ptr<IteratorBase>(new EmptyIterator(
+            {this, strings::StrCat(prefix, "::EmptySample")}));
+      } else {
+        return std::unique_ptr<IteratorBase>(
+            new Iterator({this, strings::StrCat(prefix, "::Sampling")}));
+      }
+    }
+
+    const DataTypeVector& output_dtypes() const override {
+      return input_->output_dtypes();
+    }
+
+    const std::vector<PartialTensorShape>& output_shapes() const override {
+      return input_->output_shapes();
+    }
+
+    string DebugString() const override { return "SamplingDatasetOp::Dataset"; }
+
+   protected:
+    Status AsGraphDefInternal(SerializationContext* ctx,
+                              DatasetGraphDefBuilder* b,
+                              Node** output) const override {
+      Node* input_graph_node = nullptr;
+      TF_RETURN_IF_ERROR(b->AddInputDataset(ctx, input_, &input_graph_node));
+      Node* rate = nullptr;
+      TF_RETURN_IF_ERROR(b->AddScalar(rate_, &rate));
+      TF_RETURN_IF_ERROR(b->AddDataset(this, {input_graph_node, rate}, output));
+      return Status::OK();
+    }
+
+   private:
+    class EmptyIterator : public DatasetIterator<Dataset> {
+     public:
+      explicit EmptyIterator(const Params& params)
+          : DatasetIterator<Dataset>(params) {}
+      Status GetNextInternal(IteratorContext* ctx,
+                             std::vector<Tensor>* out_tensors,
+                             bool* end_of_sequence) override {
+        *end_of_sequence = true;
+        return Status::OK();
+      }
+
+     protected:
+      std::shared_ptr<model::Node> CreateNode(
+          IteratorContext* ctx, model::Node::Args args) const override {
+        return model::MakeKnownRatioNode(std::move(args),
+                                         /*ratio=*/1);
+      }
+
+      Status SaveInternal(IteratorStateWriter* writer) override {
+        return Status::OK();
+      }
+
+      Status RestoreInternal(IteratorContext* ctx,
+                             IteratorStateReader* reader) override {
+        return Status::OK();
+      }
+    };
+
+    class Iterator : public DatasetIterator<Dataset> {
+     public:
+      explicit Iterator(const Params& params)
+          : DatasetIterator<Dataset>(params) {
+        std::random_device device("/dev/urandom");
+        std::mt19937_64 rng_temp(device());
+        rng = rng_temp;
+      }
+
+      Status Initialize(IteratorContext* ctx) override {
+        return dataset()->input_->MakeIterator(ctx, prefix(), &input_impl_);
+      }
+
+      Status GetNextInternal(IteratorContext* ctx,
+                             std::vector<Tensor>* out_tensors,
+                             bool* end_of_sequence) override {
+        // NOTE(mrry): This method is thread-safe as long as
```

Comment:
get rid of this comment

Reply:
Fixed in https://github.com/tensorflow/tensorflow/pull/25384/commits/fa11069c3e85ffbcae19994bdba09c30eee0a6de#diff-a87bcfeaacbb9697bc5998804322ce6cL129